node/src/components/small_network/tasks.rs

```diff
@@ -353,6 +362,33 @@ where
             return Err(ConnectionError::WrongNetwork(network_name));
         }
 
+        // If there is a version mismatch, we treat it as a connection error. We do not ban peers
+        // for this error, but instead rely on exponential backoff, as bans would result in issues
+        // during upgrades where nodes may have a legitimate reason for differing versions.
+        //
+        // Since we are not using SemVer for versioning, we cannot make any assumptions about
+        // compatibility, so we allow only exact version matches.
+        if context.reject_incompatible_versions
+            && protocol_version != context.chain_info.protocol_version
+        {
+            if let Some(threshold) = context.tarpit_version_threshold {
+                if protocol_version <= threshold {
+                    let mut rng = crate::new_rng();
+
+                    if rng.gen_bool(context.tarpit_chance as f64) {
+                        // If tarpitting is enabled, we hold open the connection for a specific
+                        // amount of time, to reduce load on other nodes and keep them from
+                        // reconnecting.
+                        info!(duration=?context.tarpit_duration, "tarpitting node");
```

Comment:
```suggestion
                        info!(duration = ?context.tarpit_duration, "tarpitting node");
```

Reply:
We usually do no spaces around these arguments, as far as I can tell. However, for some reason `rustfmt` reformats the `debug!` message, but not the `info!` message :angry:

Reply:
I've seen both styles; not sure whether we've agreed on any of them.
But the `tracing` docs do use spaces.